### PR TITLE
Llama: load models in HuggingFace format (incl. indexed, safetensors)

### PR DESCRIPTION
**Pull request by ochafik**
Adding flag `--model` that loads both pickled and safetensors HF formats, indexed or not, such as https://huggingface.co/meta-llama/Llama-2-7b-chat-hf or https://huggingface.co/TheBloke/Llama-2-7B-fp16

```bash
python3 examples/llama.py --prompt="Hello." --temperature=0 --gen=2 \
  --model Llama-2-7b-chat-hf/pytorch_model.bin.index.json
```

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/state.py                         23               0
------------------------------------------------------------
lines added in the tinygrad folder: 23
```

File: tinygrad/state.py [hunk 1]
```diff
@@ -45,6 +45,17 @@ def get_state_dict(obj, prefix:str='', tensor_type=Tensor) -> Dict[str, Tensor]:
 def get_parameters(obj) -> List[Tensor]: return list(get_state_dict(obj).values())
 
 def load_state_dict(model, state_dict, strict=True):
+  if 'model.embed_tokens.weight' in state_dict: # HuggingFace transformers format
+    hf_keymap = {
```
Comment (geohot) [inline tinygrad/state.py:49]: This absolutely doesn't belong here

Comment (ochafik) [reply tinygrad/state.py]: Apologies, had forgotten this wasn't llama-specific, moved to Llama.build

**Comment by geohot** [comment]
If you are changing state.py, I don't think it's right

File: tinygrad/state.py (hunk 1 above)
Comment (ochafik) [reply tinygrad/state.py]: Reverted this file, it's now all in llama.py

File: examples/llama.py [hunk 2]
```diff
@@ -218,6 +218,27 @@ def convert(name) -> Tensor:
     return lazy_tensors[0].cat(*lazy_tensors[1:], dim=axis)
   return {name: convert(name) for name in {name: None for model in models for name in model}}
 
+def load(fn:str):
+  if fn.endswith('.safetensors'):
+    from safetensors import safe_open
+    loader = safe_open(fn, framework="pt", device="cpu")
```
Comment (dc-dc-dc) [inline examples/llama.py:224]: can you use `safe_load` from `tinygrad/state` ?

Comment (ochafik) [reply examples/llama.py]: done, thanks!

File: examples/llama.py [hunk 3]
```diff
@@ -4,7 +4,7 @@
 #typeguard.importhook.install_import_hook('tinygrad')
 
 from pathlib import Path
-import functools, sys, argparse, math, platform
+import functools, sys, argparse, math, platform, json, os, re
```
Comment (wozeparrot) [inline examples/llama.py:7]: `re` isn't used, neither is `math`, `platform` or `tqdm` but those weren't added in this pr

**Comment by wozeparrot** [comment]
Any update on this? there are some conflicts.

File: examples/llama.py [hunk 4]
```diff
@@ -218,6 +218,30 @@ def convert(name) -> Tensor:
     return lazy_tensors[0].cat(*lazy_tensors[1:], dim=axis)
   return {name: convert(name) for name in {name: None for model in models for name in model}}
 
+def load(fn:str):
+  if fn.endswith('.index.json'):
+    with open(fn) as fp: weight_map = json.load(fp)['weight_map']
+    parts = {n: load(f'{os.path.dirname(fn)}/{os.path.basename(n)}') for n in set(weight_map.values())}
+    return {k: parts[n][k] for k, n in weight_map.items()}
+  elif fn.endswith('.safetensors'):
+    from tinygrad.state import safe_load
```
[the remark this answers is not in the document]
Comment (ochafik) [reply examples/llama.py]: done, thanks

**Comment by ochafik** [comment]
@wozeparrot thanks for the review & ping, got distracted this week 😅

**Pull request merged**